Updating kubectl to send delete requests with orphanDependents=false if --cascade is true

nikhiljindal opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/40568 #38897

Updating kubectl to always set `DeleteOptions.orphanDependents=false` when deleting a resource with `--cascade=true`.
This is primarily for federation where we want to use server side cascading deletion.

Impact on kubernetes: kubectl will do another GET after sending a DELETE and wait till the resource is actually deleted. This can have an impact if the resource has a finalizer. kubectl will wait till the finalizer is removed and then the resource is deleted, which is the right thing to do but a notable change in behavior.

cc @caesarxuchao @lavalamp @smarterclayton @kubernetes/sig-federation-pr-reviews @kubernetes/sig-cli-pr-reviews

nikhiljindal committed:
kubectl should always set orphanDependents=false for cascading deletion

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40576)
<!-- Reviewable:end -->

nikhiljindal commented:
I am considering this a bug fix and hence am not adding any release notes.
I dont mind adding a release note if anyone thinks that it will help.

liggitt commented:
kubectl should not be doing any GETs for deletion unless required by a client side reaper when --cascade=true

liggitt commented:
Would this (for example) make namespace deletion block the CLI?

liggitt commented:
Similar issues raised in https://github.com/kubernetes/kubernetes/issues/1899

I'm not sure --cascade is the same thing as "wait for server-side finalizers"

nikhiljindal commented:
> kubectl should not be doing any GETs for deletion unless required by a client side reaper when --cascade=true

Doing a `GET` or not is an internal detail imo. What matters is what is the user expectation from `kubectl delete`. Does it guarantee that the resource was deleted if it returns success (without timing out)? If yes, then we need the GET request.

DeploymentReaper uses server side cascading deletion and waits for deployment to be deleted (does a `GET`) after issuing `DELETE` with `DeleteOptions.OrphanDependents=false`. I assume this is the pattern we will follow for other resources as we replace Reapers with server side cascading deletion.

smarterclayton commented:
A random point before i even look at the code - will a 1.5 server properly answer orphanDependencies=false in all cases?

nikhiljindal commented:
> Would this (for example) make namespace deletion block the CLI?

Yes. And I agree that we should not do that. I guess then we cannot say that kubectl guarantees that the resource is deleted when kubectl delete returns success.

nikhiljindal commented:
> A random point before i even look at the code - will a 1.5 server properly answer orphanDependencies=false in all cases?

What do you mean by "properly answers". It will ignore it for resources that do not implement ownerRef.

nikhiljindal commented:
> > Would this (for example) make namespace deletion block the CLI?
> 
> Yes. And I agree that breaks current user expectation and hence we should not do that. I guess then we cannot say that kubectl guarantees that the resource is deleted when kubectl delete returns success.

Based on this discussion, I can update the PR to remove the `GET` request at the end. Then the change will just be to send `DeleteOptions.orphanDependents=false`. Does that sound fine?

nikhiljindal commented:
Removed the change to wait after deletion.
Now the only change is to send `DeleteOptions.orphanDependents=false` when `--cascade=true`.
@liggitt @smarterclayton PTAL.

liggitt commented:
> A random point before i even look at the code - will a 1.5 server properly answer orphanDependencies=false in all cases?

did you test this against a 1.5 server with various resource types?

cc @caesarxuchao @deads2k any feedback on clients always sending this?

nikhiljindal commented:
@k8s-bot gce etcd3 e2e test this

caesarxuchao commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -251,7 +251,8 @@ func ReapResult(r *resource.Result, f cmdutil.Factory, out io.Writer, isDefaultD
 			if kubectl.IsNoSuchReaperError(err) && isDefaultDelete {
-				return deleteResource(info, out, shortOutput, mapper)
+				// No client side reaper found. Let the server do cascading deletion.
+				return cascadingDeleteResource(info, out, shortOutput, mapper)
```
Is it only the federation resources that are missing reapers? Is it possible that users are relying on kubectl not cascading delete federation resources?

nikhiljindal replied:
> Is it only the federation resources that are missing reapers?

Kubernetes resources are also missing reapers for resources where there is nothing to reap (for ex: secrets, configmaps). In federation, we need to reap them from underlying clusters. This change should not have any impact on any kubernetes resource.

> Is it possible that users are relying on kubectl not cascading delete federation resources?

Cascading deletion should happen when users use `kubectl delete --cascading=true` with federation. Right now, it cascading deletes some resources (like namespaces, rs and deployments) and does not cascading deletes others resources (like secrets, configmaps). This is a bug that this PR fixes.

Default behavior is being discussed in https://github.com/kubernetes/kubernetes/issues/38897. The consensus seems to be to be consistent with kubernetes (i.e cascading delete by default).

caesarxuchao replied:
Thank you for the explanation. I don't have further questions. @liggitt @smarterclayton do you have more comments?

smarterclayton replied:
We should probably be cascading even for things that have reapers (reaper is a client implementation of cascade, so the intent is to cascade).  Do all existing reapers cascade?  If no, they should also orphan: false.

nikhiljindal replied:
Done

nikhiljindal commented:
cc @smarterclayton Can you please take a look?

smarterclayton commented on pkg/kubectl/cmd/delete.go:
```diff
@@ -302,8 +303,14 @@ func DeleteResult(r *resource.Result, out io.Writer, ignoreNotFound bool, shortO
 
-func deleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper) error {
-	if err := resource.NewHelper(info.Client, info.Mapping).Delete(info.Namespace, info.Name); err != nil {
+func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper) error {
```
Add a test case.

nikhiljindal replied:
Done

smarterclayton commented:
Can you verify against 1.5 that kubectl still works?  Also, would like to see an e2e test that covers this now that we're fundamentally changing e2e.  With that and the comments above addressed I'd be ok with this change.

smarterclayton reviewed (changes requested):
Comment

nikhiljindal commented:
>  Also, would like to see an e2e test that covers this now that we're fundamentally changing e2e

Sorry not sure what you are referring to in "now that we're fundamentally changing e2e". Can you please elaborate?
All existing kubectl tests (make test-cmd) that run `kubectl delete` execute this code path already.

nikhiljindal committed:
Updating reapers to set OrphanDependents=false

nikhiljindal committed:
Adding a unit test for verifying OrphanDependents in kubectl delete requests

nikhiljindal commented:
Rebased and updated the PR as per comments.
Updated all reapers to set OrphanDependents=false and added a unit test.
Also verified that this works with 1.5 apiserver.

@smarterclayton PTAL.

smarterclayton commented:
> Also, would like to see an e2e test that covers this now that we're fundamentally changing e2e

Re test-cmd it doesn't stress GC in a meaningful way at least when I'm looking.  You're changing how GC works from client reaper, I'd like to see a test that passes before and after that ensures this change works as designed.

nikhiljindal committed:
Add cascading deletion test cases to kubectl tests

nikhiljindal commented:
@smarterclayton Pushed a new commit with more cascading deletion test cases for deployment and rs.
For non controllers (configmaps, secrets), we already have simple deletion tests.
Let me know if this is not what you were expecting.

nikhiljindal commented:
@k8s-bot unit test this
@k8s-bot verify test this

nikhiljindal commented:
Fixed the failing unit test and rebased.
@smarterclayton Can I have LGTM and approval?

smarterclayton commented:
/approve
/lgtm

thanks, tests are good

nikhiljindal commented:
@k8s-bot kubemark e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot node e2e test this
@k8s-bot gce e2e test this
@k8s-bot gce etcd3 e2e test this
